[#68226] show formatted weights in workpackages

Kharonus opened the pull request:
# Ticket
[#68226](https://community.openproject.org/wp/68226)

# What are you trying to accomplish?
- show weights in display fields and select edit fields

# What approach did you choose and why?
- use item formatter for APIv3 responses

Kharonus committed:
[#68226] show formatted weights in workpackages

- https://community.openproject.org/wp/68226
- use item formatter for APIv3 responses
- show weights in display fields and select edit fields

Kharonus committed:
[#68226] fixed representer spec

NobodysNightmare commented on frontend/src/app/shared/components/fields/display/field-types/render-hierarchy-item.ts:
```diff
@@ -68,6 +68,11 @@ function spansFromAncestors(ancestors:CollectionResource):HTMLSpanElement[] {
         spans.push(short);
+      } else if (el.weight !== null) {
+        const weight = document.createElement('span');
+        weight.textContent = `(${el.weight})`;
```
:question: Do I see it right, that this weight is not formatted through our universal number formatter?

Kharonus replied:
The `weight` property here is returned from the API and should be already formatted. We have no access to our formatting logic from the frontend code.

Kharonus replied:
this now uses the formatted weight

NobodysNightmare commented on lib/api/v3/custom_fields/hierarchy/hierarchical_item_aggregate.rb:
```diff
@@ -40,9 +40,30 @@ def initialize(item:, depth:)
           end
 
+          def weight
+            weight_formatter.format(item:)
```
:yellow_circle:  Mh. I understand that we need to decide on an opinionated default representation for the `title`. However, for the weight I don't quite understand that we need to do this.

One first note: Doesn't the usage of the formatter lead to a weight of `nil` being represented as `""`? (and do we have a spec for `""` vs `nil`?)

But even that aside: Shouldn't we be representing the weight as a number, which essentially means a float in JSON? I'd expect the API client to care about formatting the number or using it in full precision for calculations.

Kharonus replied:
That is a valid take, but none I can easily answer.

We have not the same access to formatting in the OUR API client, meaning the frontend. right now I exposed the weight as already formatted represented value.

But I get the point, that we might want to actually deliver the "correct" number here. But at the same time, we then probably do not want to use float, because we would loose precision.

Kharonus replied:
there is another option. For the API we could return both, the precise number and a formatted representation.

NobodysNightmare replied:
> But at the same time, we then probably do not want to use float, because we would loose precision.

Yeah, that's a tricky point. I am pretty sure that `BigDecimal#to_json` returns a string for that reason, though the more portable way would be to return a float.

If we are concerned about our API client, we could expose `weight` and `formatted_weight`. What do you think?

Kharonus replied:
I added the code to return both, weight and formatted weight

Kharonus committed:
[#68226] use formatted weight for hierarchy items

- hierarachy item now return a weight and a formatted weight in the API
  response
- weight is the accurate value
- formatted weight is the value in display representation after standard
  formatting

NobodysNightmare commented on docs/api/apiv3/components/schemas/hierarchy_item_read_model.yml:
```diff
@@ -13,8 +13,29 @@ properties:
-    description: The short name of the hierarchy item
+    type:
+      - string
+      - "null"
```
:green_circle: This is already very good, just a side-note because I am currently working on schemas: Every key that's never going to be skipped, should be marked as required. I.e. since `short` and `weight` are `null` if they do not apply, they seem to always be rendered. This means we should mark them as required (and probably more keys).

Kharonus replied:
Okay, will do.

Kharonus committed:
[#68226] refine api item read model
